src/mono/wasm/Marshalling.md

```diff
@@ -0,0 +1,104 @@
+# State of method binding and parameter marshalling as of 2022-01-17
+This is not 100% exhaustive. This is not intended as guide to exernal usage, rather as inventory before cleanup.
+
+## Calling from JS to C
+- imported functions (WASM)
+  - is handle of C method, which could be called from JS with numeric parameters
+  - see `EMSCRIPTEN_KEEPALIVE` macro in C code
+- `ccall` (emscripten)
+  - helper to make call to C method pointer (export)
+  - we use it for all calls from JS
+  - converts JS strings to utf8 (char*)
+  - copy byte array to WASM memory
+  - converts return to boolean
+  - allocates parameters on WASM stack!
+  - we use it to implement `mono_wasm_invoke_method` and `mono_wasm_get_delegate_invoke`
+- `cwrap` (emscripten)
+  - create JS proxy to C method
+  - are emscripten javascript proxies for calling existing C function
+  - they use `ccall`
+  - used in `cwraps.ts`
+
+## Calling from JS to C#
+- `mono_wasm_invoke_method` (JS runtime)
+  - is JS proxy of C `mono_runtime_invoke`, via `cwrap` of `ccall`
+  - which is way how to call C# from C
+- `call_method` (JS runtime)
+  - is way how to call C# method from JS
+  - generates parameter converters and uses them
+  - calls to C# via `mono_wasm_invoke_method`, via `ccall` to C `mono_runtime_invoke`
+- `mono_bind_method` (JS runtime)
+  - generates JS proxy for C# method
+  - generates parameter converters
+    - allocates parameters on WASM heap, C# or primitive types
```

Comment:
where possible it reuses an existing heap buffer to avoid an allocation per call, but that's an optimization

Reply:
Updated, thanks